### DISCUSSION TITLE
Performance Issue on Firefox and Safari on Desktop

### DISCUSSION THREAD

**Discussion by jn-cavera** [alpinejs/alpine#3153] [7. Bugs (archived)]
I have built an application that creates customized products based on a users selection (like Dell's computer builder basically). The code is attached.

It works flawlessly on most browsers but after doing some testing, am seeing some issues on Desktop versions of Firefox and Safari. When using Firefox, I'm finding a very long delay between my function completing and the changes being reflected on screen. I don't believe it's an issue with the changeAddon() function because even if I remove all code from that function, the issue remains.

In Safari, the browser actually doesn't load the page at all and just locks up.

I do have multiple versions of this loop running on my product pages and some of the loops contain a lot of product options (like 50 products each with multiple options themselves). I'm guessing it has to do with how these two browsers handle the larger DOM sizes but not sure how to fix it. Also, things seem to work fine on the iOS versions of Safari. It's only the desktop version which really makes no sense to me.
[code-snippet.txt](https://github.com/alpinejs/alpine/files/9543359/code-snippet.txt)

Would love to hear if anyone out there is having similar problems and if they've been able to figure out a work-around.

**Comment by ekwoka** [comment]
This shouldn't really be a problem. Benchmarks and such run thousands of rows for loops, and I've done some things that were slightly similar.

One thing I'd recommend though is to reduce how much code is in the expression and offload more of those to the data component.

It can make refactoring much easier.

But one thing that I can see that is wrong is that you don't have a `:key` on the for loops. That can have performance issues as Alpine basically has to rerender the entire loop every time.
